packages/nextjs-mf/src/plugins/NextFederationPlugin.ts

```diff
@@ -38,187 +52,470 @@ const regexEqual = (x, y) => {
+  };
+}
 
-  constructor(options: NextFederationPluginOptions) {
-    const { extraOptions, ...mainOpts } = options;
-    this._options = mainOpts;
-
-    this._extraOptions = {
-      automaticPageStitching: false,
-      enableImageLoaderFix: false,
-      enableUrlLoaderFix: false,
-      skipSharingNextInternals: false,
-      automaticAsyncBoundary: false,
-      ...extraOptions,
+// Utility function to validate compiler options
+function validateCompilerOptions(compiler: Compiler): boolean {
+  if (!compiler.options.name) {
+    throw new Error('name is not defined in Compiler options');
+  }
+
+  if (!['server', 'client'].includes(compiler.options.name)) {
+    return false;
+  }
+  return true;
+}
+
+// Utility function to validate NextFederationPlugin options
+function validatePluginOptions(options: ModuleFederationPluginOptions): void {
+  if (!options.filename) {
+    throw new Error('filename is not defined in NextFederation options');
+  }
+}
+
+/**
+ * Configures server-specific compiler options.
+ * @param {Compiler} compiler - The Webpack compiler instance.
+ */
+function configureServerCompilerOptions(compiler: Compiler): void {
+  compiler.options.target = false;
+  compiler.options.optimization.chunkIds = 'named';
+  compiler.options.optimization.splitChunks = false;
+}
+
+/**
+ * Applies server-specific plugins.
+ * @param {Compiler} compiler - The Webpack compiler instance.
+ * @param options
+ */
+function applyServerPlugins(
+  compiler: Compiler,
+  options: ModuleFederationPluginOptions
+): void {
+  const { StreamingTargetPlugin } = require('@module-federation/node');
+
+  new AddModulesPlugin({
+    runtime: 'webpack-runtime',
+    eager: true,
+    remotes: options.remotes,
+  }).apply(compiler);
+
+  new AddModulesPlugin({
+    runtime: options.name,
+    eager: false,
+    remotes: options.remotes,
+  }).apply(compiler);
+
+  new StreamingTargetPlugin(options, {
+    ModuleFederationPlugin: compiler.webpack.container.ModuleFederationPlugin,
+  }).apply(compiler);
+}
+
+/**
+ * Configures server-specific library and filename options.
+ * @param {ModuleFederationPluginOptions} options - The ModuleFederationPluginOptions instance.
+ */
+function configureServerLibraryAndFilename(
+  options: ModuleFederationPluginOptions
+): void {
+  options.library = {
+    type: 'commonjs-module',
+    name: options.name,
+  };
+  options.filename = path.basename(options.filename as string);
+}
+
+/**
+ * Handles server-specific externals.
+ * @param {Compiler} compiler - The Webpack compiler instance.
+ * @param {ModuleFederationPluginOptions} options - The ModuleFederationPluginOptions instance.
+ */
+function handleServerExternals(
+  compiler: Compiler,
+  options: ModuleFederationPluginOptions
+): void {
+  if (Array.isArray(compiler.options.externals)) {
+    const originalExternals = compiler.options.externals[0];
+
+    compiler.options.externals[0] = async function (ctx, callback) {
+      if (
```

Comment:
May be add a comment here describing what this is doing? Its kinda hard to understand what is happening here.

Reply:
done in next pr